src/dmd/link.d

```diff
@@ -569,36 +569,23 @@ public int runLINK()
             argv.push("-Xlinker");
             argv.push("--gc-sections");
         }
-        /* Add each library, prefixing it with "-l".
+        /* Add each library, prefixing it with "-l" if necessary.
          * The order of libraries passed is:
-         *  1. any libraries passed with -L command line switch
-         *  2. libraries specified on the command line
+         *  1. static libraries ending with *.a
+         *  2. link switches passed with -L command line switch
          *  3. libraries specified by pragma(lib), which were appended
          *     to global.params.libfiles.
-         *  4. link switches, that may also contain -l libraries
-         *  5. standard libraries.
+         *  4. standard libraries.
          */
-        for (size_t i = 0; i < global.params.libfiles.dim; i++)
+        foreach (i; 0 .. global.params.libfiles.dim)
         {
             const(char)* p = global.params.libfiles[i];
-            size_t plen = strlen(p);
-            if (plen > 2 && p[plen - 2] == '.' && p[plen - 1] == 'a')
+            const plen = strlen(p);
+            const isArchive = (plen > 2 && p[plen - 2] == '.' && p[plen - 1] == 'a');
```

Comment:
`dmd.root.filename` has utilities to deal with filenames.
From the top of my head, `const isArchive = FileName.ext(p) == "a";` might be way more readable.